src/or/hs_descriptor.c

```diff
@@ -234,53 +234,69 @@ build_mac(const uint8_t *mac_key, size_t mac_key_len,
   crypto_digest_free(digest);
 }
 
-/* Using a given decriptor object, build the secret input needed for the
- * KDF and put it in the dst pointer which is an already allocated buffer
+/* Using a secret data and a given decriptor object, build the secret
+ * input needed for the KDF.
+ *
+ * secret_input = SECRET_DATA | subcredential | INT_8(revision_counter)
+ *
+ * Then, put it in the dst pointer which is an already allocated buffer
  * of size dstlen. */
 static void
-build_secret_input(const hs_descriptor_t *desc, uint8_t *dst, size_t dstlen)
+build_secret_input(const hs_descriptor_t *desc,
+                   const uint8_t *secret_data,
+                   size_t secret_data_len,
+                   uint8_t *dst, size_t dstlen)
 {
   size_t offset = 0;
+  size_t outlen = secret_data_len + DIGEST256_LEN + sizeof(uint64_t);
 
   tor_assert(desc);
+  tor_assert(secret_data);
   tor_assert(dst);
-  tor_assert(HS_DESC_ENCRYPTED_SECRET_INPUT_LEN <= dstlen);
+  tor_assert(outlen <= dstlen);
 
   /* XXX use the destination length as the memcpy length */
-  /* Copy blinded public key. */
-  memcpy(dst, desc->plaintext_data.blinded_pubkey.pubkey,
-         sizeof(desc->plaintext_data.blinded_pubkey.pubkey));
-  offset += sizeof(desc->plaintext_data.blinded_pubkey.pubkey);
+  /* Copy the secret data. */
+  memcpy(dst, secret_data, secret_data_len);
+  offset += secret_data_len;
   /* Copy subcredential. */
   memcpy(dst + offset, desc->subcredential, sizeof(desc->subcredential));
   offset += sizeof(desc->subcredential);
   /* Copy revision counter value. */
   set_uint64(dst + offset, tor_htonll(desc->plaintext_data.revision_counter));
   offset += sizeof(uint64_t);
-  tor_assert(HS_DESC_ENCRYPTED_SECRET_INPUT_LEN == offset);
+  tor_assert(outlen == offset);
 }
 
 /* Do the KDF construction and put the resulting data in key_out which is of
  * key_out_len length. It uses SHAKE-256 as specified in the spec. */
 static void
 build_kdf_key(const hs_descriptor_t *desc,
+              const uint8_t *secret_data,
+              size_t secret_data_len,
               const uint8_t *salt, size_t salt_len,
               uint8_t *key_out, size_t key_out_len,
               int is_superencrypted_layer)
 {
-  uint8_t secret_input[HS_DESC_ENCRYPTED_SECRET_INPUT_LEN];
+  uint8_t *secret_input = NULL;
+  size_t secret_input_len;
   crypto_xof_t *xof;
 
   tor_assert(desc);
+  tor_assert(secret_data);
   tor_assert(salt);
   tor_assert(key_out);
 
+  secret_input_len = secret_data_len + DIGEST256_LEN + sizeof(uint64_t);
```

Comment:
Hmm, `build_secret_input()` has a bit weird interface now that it receives an already allocated buffer and puts stuff in it. I think perhaps a more clean way to do this would be to have `build_secret_input()` allocate the buffer and return it. That means you wont have to compare `dstlen` with `outlen` etc.

If you do that, `build_secret_input()` will not be able to fail, so you don't need to check its output for NULL. You can just assert it if you want.